application-analytics-ui/src/main/resources/Analytics/Code/Macros/RowEvolution.xml

```diff
@@ -0,0 +1,289 @@
+          fill: false
+        }]
+      },
+      options: {
+        scales: {
+          y: {
+            beginAtZero: true
+          }
+        }
+      }
+    });
+    $('#chartModal').modal('show');
+  };
+
+  /**
+   * Fetches the data from Matomo using the XWiki server as a proxy through the RowEvolutionJSON page. It then saves the data in session 
+   * storage for easy access. This allows the user to display a different field for the same time interval without making a new request and updates the chart.
+   *
+   * @param parameter parameter need for the request
+   */
+  const fetchChartData = (parameter) =&gt; {
+    $('#options').data('parameter', parameter);
+    const options = $('#options')
+    const period = options.data('period')
+    const date = options.data('date');
+    const requestPage = options.data('requestPage')
+    $.getJSON(`${requestPage}?parameter=${parameter}&amp;period=${period}&amp;date=last${date}`)
+      .done((data) =&gt; {
+        window.sessionStorage.setItem('analyticsLastRequestCache', JSON.stringify(data));
+        renderChart(data);
+      })
+      .fail((jqxhr, textStatus, error) =&gt; {
+        const err = `${textStatus}, ${error}`;
+        console.error(`Request Failed: ${err}`);
+      });
+  };
+
+  /**
+   * Will show only the time options available for the period given as a parameter and updates the options
+   * 
+   * @param period Day/Week/Month/Year
+   */
+  const displayPeriodDurations = (period) =&gt; {
+    PERIODS.forEach(p =&gt; {
+      p === period ? $(`#periodDuration${p}`).show() : $(`#periodDuration${p}`).hide();
+    });
+    $('#options').data('period', period.toLowerCase());
+    $('#options').data('date', $(`#periodDuration${period}`).val());
+  };
+
+  /**
+   * Resets to the default option for all the selectos when the modal is closed.
+   */
+  const restOptions = () =&gt; {
+    $('#chartModal').on('hidden.bs.modal', function () {
+      PERIODS.forEach(p =&gt; {
+        $(`#periodDuration${p}`).prop('selectedIndex', 0);
+      });
+      $('#field').prop('selectedIndex', 0);
+    });
+  };
+
+  const displayStatisitcs = () =&gt; {
+    $('#field').on('change', function () {
+      renderChart(JSON.parse(window.sessionStorage.getItem('analyticsLastRequestCache')));
+    });
+  };
+
+  /**
+   * Binds the event that will open the modal and display the data when the 'Row Evolution' button is clicked. 
+   */
+  const bindRowEvolutionDisplay = () =&gt; {
+    $(document).on('click', '.displayMostViewedPagesActions', function () {
+      displayPeriodDurations('Day');
+      const parameter = $(this).data('parameter');
+      $('#periodSelect').prop('selectedIndex', 0);
+      $('#options').data('date', '30');
+      fetchChartData(parameter);
+    });
+  };
+  
+  /**
+   * Updates the chart when the time  
+   */
+  const updateOnPeriodChange = () =&gt; {
+    PERIODS.forEach(p =&gt; {
+      $(`#periodDuration${p}`).on('change', function () {
```

Comment:
Here as well, you can have outside of this method a `$(document).on('keyup change', 'mySelector', function() {...`
But with a better selector, since you cannot use ids for `#period...`, since this macro can be added multiple times on a page. 
You also don't need a different listener for each period. You can simply listen for changes on elements that have a class that starts with `periodDuration`